src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -433,192 +395,163 @@ createBlockHistoryTable =
+      assign bsBlockHeight bRestore
+      let msg = "handlePossibleRewind: newChildBlock: error finding txid"
+      SInt txid <- callDb "getting txid" $ \db ->
+        expectSingleRowCol msg =<< qry db
             "SELECT endingtxid FROM BlockHistory WHERE blockheight = ?;"
             [SInt (fromIntegral bCurrent)]
             [RInt]
-        bs@(BlockVersion bh version) <- gets _bsBlockVersion
-        result <- callDb "handleVersion" $ \db -> qry db
-            "SELECT txid FROM VersionHistory WHERE blockheight = ? AND version = ?;"
-            [SInt (fromIntegral bh), SInt (fromIntegral version)]
-            [RInt]
-        newtxid <- if null result
-            then do
-                versionHistoryInsert bs (fromIntegral txid)
-                return $! (fromIntegral txid :: TxId)
-            else do
-               r <- liftIO $
-                    expectSingleRowCol "handleVersion: block version check"
-                        result
-               case r of
-                   SInt res -> return $! (fromIntegral res)
-                   _ -> error "handleVersion: We failed to get an integer for \
-                              \the txid."
-        assign bsTxId newtxid
-        return newtxid
-
-    rewindBlock vCurrent = do
-        bsBlockVersion .= BlockVersion bRestore (succ vCurrent)
-        tableMaintenanceRowsVersionedTables
-        dropUserTables
-        t <- deleteHistory
-        assign bsTxId t
-        b <- gets _bsBlockVersion
-        versionHistoryInsert b t
-        return $! t
-
-tableMaintenanceRowsVersionedTables :: BlockHandler SQLiteEnv ()
-tableMaintenanceRowsVersionedTables = do
-    tableMaintenanceRowsVersionedSystemTables
-    tableMaintenanceRowsVersionedUserTables
+      assign bsTxId (fromIntegral txid)
+      return $ fromIntegral txid
+
+    rewindBlock bh = do
+      assign bsBlockHeight bh
+      tableMaintenanceRowsVersionedSystemTables
+      callDb "rewindBlock" $ \db -> do
+          droppedtbls <- dropTablesAtRewind bh db
+          vacuumTablesAtRewind bh droppedtbls db
+      t <- deleteHistory
+      assign bsTxId t
+      return $! t
+
+dropTablesAtRewind :: BlockHeight -> Database -> IO [Utf8]
+dropTablesAtRewind bh db = do
+  toDropTblNames <- qry db "SELECT tablename FROM VersionedTableCreation where createBlockheight >= ?;"
+    [SInt (fromIntegral bh)]
+    [RText]
+  forM toDropTblNames $ \case
+    [SText tblname] -> do
+      exec_ db $ "DROP TABLE " <> tblname <> ";"
+      return tblname
+    _ -> internalError "rewindBlock: dropTablesAtRewind: Couldn't resolve the name of the table to drop."
+
+vacuumTablesAtRewind :: BlockHeight -> [Utf8] -> Database -> IO ()
+vacuumTablesAtRewind bh droppedtbls db = do
+  toVacuumTblNames <- qry db "SELECT DISTINCT tablename FROM VersionedTableMutation WHERE blockheight >= ?;"
+    [SInt (fromIntegral bh)]
+    [RText]
+  forM_ toVacuumTblNames $ \case
+    [SText tblname] -> do
+      when (notElem tblname droppedtbls) $
```

Comment:
`notElem` on list is `O(n)` and this will be called a lot. You have two options for what to return from `dropTablesAtRewind` -- a sorted `Vector` (which you can `bsearch` on using [vector-algorithms](http://hackage.haskell.org/package/vector-algorithms-0.8.0.1/docs/Data-Vector-Algorithms-Search.html)), or a `HashSet Utf8`. `HashSet` gives you the benefit of allowing you to just compute a set difference here and then you don't need the `when` condition.

Reply:
done